pkg/spdk/replica.go

```diff
@@ -1543,6 +1609,558 @@ func (r *Replica) SnapshotRangeHashGet(spdkClient *spdkclient.Client, snapshotNa
+			setStatus(status.State, status.ErrorMsg, status.ProcessedClusters, status.TotalClusters)
+
+			if status.State == types.ProgressStateError || status.State == types.ProgressStateComplete {
+				r.log.Infof("Clone state: %v", status.State)
+				return
+			}
+		}
+	}
+}
+
+func (r *Replica) SnapshotCloneDstStatusCheck() (status *spdkrpc.ReplicaSnapshotCloneDstStatusCheckResponse, err error) {
+	r.Lock()
+	defer r.Unlock()
+
+	defer func() {
+		err = errors.Wrapf(err, "failed to check snapshot clone status in dst replica %v", r.Name)
+	}()
+
+	c := r.snapshotCloningDstCache
+	var progress uint32
+	if c.totalClusters > 0 {
+		progress = uint32(c.processedClusters / c.totalClusters * 100)
+	}
+
+	if c.cloningState == types.ProgressStateComplete {
+		progress = 100
+	}
+
+	return &spdkrpc.ReplicaSnapshotCloneDstStatusCheckResponse{
+		IsCloning:         r.isSnapshotCloning,
+		SrcReplicaName:    c.srcReplicaName,
+		SrcReplicaAddress: c.srcReplicaAddress,
+		SnapshotName:      c.snapshotName,
+		State:             c.cloningState,
+		Progress:          progress,
+		Error:             c.cloningError,
+	}, nil
+}
+
+func (r *Replica) SnapshotCloneDstFinish(spdkClient *spdkclient.Client, cloneMode spdkrpc.CloneMode) (err error) {
+	if cloneMode == spdkrpc.CloneMode_CLONE_MODE_LINKED_CLONE {
+		r.isSnapshotCloning = true
+		return nil
+	}
+
+	updateRequired := false
+
+	r.Lock()
+	defer func() {
+		r.Unlock()
+
+		if updateRequired {
+			r.UpdateCh <- nil
+		}
+	}()
+
+	if !r.isSnapshotCloning {
+		return fmt.Errorf("replica %s is not in cloning", r.Name)
+	}
+
+	defer func() {
+		if err != nil {
+			if r.State != types.InstanceStateError {
+				r.State = types.InstanceStateError
+			}
+			r.ErrorMsg = err.Error()
+			if r.snapshotCloningDstCache.cloningError == "" {
+				r.snapshotCloningDstCache.cloningError = err.Error()
+				r.snapshotCloningDstCache.cloningState = types.ProgressStateError
+			}
+		} else {
+			if r.State != types.InstanceStateError {
+				r.ErrorMsg = ""
+			}
+		}
+
+		updateRequired = true
+	}()
+
+	if r.snapshotCloningDstCache.cloningState == types.ProgressStateComplete {
+		if r.Head == nil {
+			return fmt.Errorf("cannot find the head for replica %s snapshot clone finish", r.Name)
+		}
+		if r.snapshotCloningDstCache.cloningLvol == nil {
+			return fmt.Errorf("cannot find the head for cloning lvol for snapshot clone finish in replica %v", r.Name)
+		}
+		tmpSnapName := GetTmpSnapNameForCloningLvol(r.Name)
+		snapUUID, err := spdkClient.BdevLvolSnapshot(r.snapshotCloningDstCache.cloningLvol.UUID, tmpSnapName, []spdkclient.Xattr{})
+		if err != nil {
+			return err
+		}
+		snapBdevLvol, err := spdkClient.BdevLvolGetByName(snapUUID, 0)
+		if err != nil {
+			return err
+		}
+		tmpSnap := BdevLvolInfoToServiceLvol(&snapBdevLvol)
+		if _, err := spdkClient.BdevLvolSetParent(r.Head.Alias, tmpSnap.Alias); err != nil {
+			return err
+		}
+	}
+
+	if err = r.doCleanupForSnapshotCloneDst(spdkClient, false); err != nil {
+		return err
+	}
```

Comment:
Need to try its best to execute this cleanup even if there is an error above.

Reply:
Here the replica is marked as error then the cleanup will be handled when Longhorn manager stop this replica at https://github.com/longhorn/longhorn-spdk-engine/blob/9413f9effe18ad7f31bae1f9d36f00253d9f3226/pkg/spdk/replica.go#L1049 

WDYT?